### PR TITLE
Language support policy

### PR DESCRIPTION
**Pull request by pkooij**
## Summary

This PR adds language-conditioned policy training and interactive deployment to LeRobot.

It introduces **PI052**, a self-contained π0.5 variant that keeps the continuous action expert while re-enabling PaliGemma text generation. PI052 can be trained from recipe-rendered language annotations to predict low-level subtasks, update memory, answer visual questions, and condition its action chunks on the active subtask.

The deployment side is integrated into **`lerobot-rollout`**. The same runtime can:

- let PI052 generate its own active subtask from a high-level task;
- bypass the high-level planner and let the operator supply each subtask directly;
- run in a RoboCasa simulation or on a real robot;
- run flat language-conditioned policies such as PI0.5 and MolmoAct2 in direct-task mode.

Related work:

- π0.5: [arxiv:2504.16054](https://arxiv.org/abs/2504.16054)
- Language annotation pipeline: #3471

### PI052 policy and training

- Adds `PI052Config`, `PI052Policy`, processors, tokenizer support, and the runtime adapter under `src/lerobot/policies/pi052/`.
- Trains configurable combinations of:
  - continuous flow-matching action loss;
  - language-model cross-entropy on supervised language turns;
  - optional FAST-tokenized action cross-entropy.
- Adds selective text/FAST supervision, attention-mask handling, autoregressive `select_message`, KV-cached decoding, prompt-component dropout, and `<locDDDD>` token support.
- Supports checkpoint-safe PI052 processor reconstruction and restoration of saved normalization state.
- Includes optional FAST tokenizer fitting and optional FlashRT FP8 MLP inference support.

### Recipe-driven language data

Adds three training recipes:

- `recipes/subtask.yaml`: subtask prediction + subtask-conditioned actions;
- `recipes/subtask_mem.yaml`: adds persistent memory updates;
- `recipes/subtask_mem_vqa_speech.yaml`: adds VQA and spoken interjection supervision.

`RenderMessagesStep`, language rendering, collation, and PI052 tokenization turn the annotation schema into the appropriate high-level text targets and low-level action prompts. Missing optional annotation streams are skipped through recipe `if_present` bindings.

### Interactive language runtime

- Integrates language deployment into `lerobot-rollout`.
- Adds a policy-adapter registry for PI052, PI0.5, and MolmoAct2.
- Reuses the standard rollout context for real robots, including Draccus robot configuration, feature schemas, policy and robot processors, device configuration, safety processing, and action dispatch.
- Accepts registered `--robot.*` and `--policy.*` configuration fields instead of maintaining a separate robot configuration surface.
- Supports live task replacement, direct-subtask mode, pause/resume, action-chunk queues, configurable language generation, optional memory, and Rerun visualization.

### RoboCasa and evaluation

- Adds a persistent RoboCasa backend with multi-view rendering, live MJPEG streaming, interactive prompts, scene reset, and annotated MP4 output.
- Adds task/subtask overlays to evaluation videos using the same single-pass annotation helper as the simulator.
- Extends RoboCasa environment configuration used by evaluation and the interactive runtime.

### Training integration

- Adds PI052 policy factory and processor registration.
- Adds training metrics and diagnostics for the individual loss components and generated text samples.
- Includes optimizer and distributed-training adjustments used for large PI05/PI052 fine-tuning.

### Installation

```bash
git clone https://github.com/huggingface/lerobot.git
cd lerobot
python -m venv .venv
source .venv/bin/activate
pip install -e ".[pi]"
```

RoboCasa is installed separately; follow the [RoboCasa installation guide](https://robocasa.ai/) before using `--sim`. Real-robot runs also require the dependencies for the selected robot and cameras.

### PI052 in RoboCasa

```bash
export MUJOCO_GL=egl

lerobot-rollout \
  --policy.path=lerobot/pi052_robocasa \
  --sim \
  --sim.task=CloseFridge \
  --sim.split=pretrain \
  --task="close the fridge" \
  --disable_memory \
  --sim.render_size=384 \
  --sim.views=robot0_agentview_left,robot0_eye_in_hand,robot0_agentview_right \
  --mode=action \
  --ctrl_hz=20
```

Open <http://localhost:8010> for the live view. Use `/pause`, `/resume`, and `/reset`, or enter a new prompt at any time.

Without `--sim.direct_subtask`, the prompt is the high-level task and PI052 generates the active low-level subtask. Add `--sim.direct_subtask` to bypass that planner and send each prompt directly to the action policy.

### MolmoAct2 on an SO-100/SO-101

MolmoAct2 is a low-level language-conditioned policy and does not contain the MolmoAct2-ER high-level planner. Run it in direct-subtask mode:

```bash
lerobot-rollout \
  --policy.path=lerobot/MolmoAct2-SO100_101-LeRobot \
  --policy.device=cuda \
  --robot.type=so101_follower \
  --robot.port=/dev/ttyACM1 \
  --robot.cameras='{"cam0":{"type":"opencv","index_or_path":"/dev/video0","width":640,"height":480,"fps":30,"fourcc":"MJPG","backend":200},"cam1":{"type":"opencv","index_or_path":"/dev/video2","width":640,"height":480,"fps":30,"fourcc":"MJPG","backend":200}}' \
  --direct_subtask \
  --rerun \
  --robot.max_relative_target='{"shoulder_pan":5,"shoulder_lift":5,"elbow_flex":5,"wrist_flex":5,"wrist_roll":5,"gripper":5}'
```

With no initial `--task`, the robot starts paused. Check the workspace and configured motion limits before entering the first subtask. A new prompt replaces the current instruction; `/pause` and `/resume` control action dispatch.

## Validation

- `pytest -p no:cacheprovider tests/runtime tests/configs/test_recipe.py tests/datasets/test_language_render.py tests/processor/test_render_messages_processor.py tests/policies/pi052 tests/test_rollout.py -q`
  - **96 passed, 2 skipped**
- Pre-commit passes for the latest runtime consolidation, including Ruff, Bandit, mypy, formatting, and typos.
- Manual validation has covered RoboCasa execution and SO-100/SO-101 language-conditioned rollouts.

## Notes and limitations

- The language runtime is experimental; policies still make mistakes and operators remain responsible for robot safety.
- PI052 provides high-level text generation. PI0.5 and MolmoAct2 use the shared direct-task adapter and should be run with `--direct_subtask` (or `--sim.direct_subtask`).
- Standard non-language rollout strategies continue to use the existing `lerobot-rollout` path. Language mode is selected explicitly with `--language` or automatically by language-only flags such as `--sim` and `--direct_subtask`.

## Checklist

- [x] Targeted policy, processor, recipe, dataset, runtime, and rollout tests pass locally
- [x] Runtime consolidation passes pre-commit
- [x] Documentation updated
- [ ] Full PR CI is green
- [ ] Community review: link another reviewed contributor PR

### CODE REVIEW

**Comment by pkooij** [comment]
@claude review

**Comment by pkooij** [comment]
Superseded by the focused stacked split: #4182 → #4183 → #4184 → #4056 → #4076. The existing RTC and MEM branches were rebuilt on top of the new PI052 tip and their PR bases were retargeted, so the stack no longer carries this PR’s 404-commit ancestry.
